src/sentry/api/paginator.py

```diff
@@ -220,3 +224,101 @@ def get_result(self, limit=100, cursor=None):
             next=next_cursor,
             prev=prev_cursor,
         )
+
+
+def reverse_bisect_left(a, x, lo=0, hi=None):
+    """\
+    Similar to ``bisect.bisect_left``, but expects the data in the array ``a``
+    to be provided in descending order, rather than the ascending order assumed
+    by ``bisect_left``.
+
+    The returned index ``i`` partitions the array ``a`` into two halves so that:
+
+    - left side: ``all(val > x for val in a[lo:i])``
+    - right side: ``all(val <= x for val in a[i:hi])``
+    """
+    if lo < 0:
+        raise ValueError('lo must be non-negative')
+
+    if hi is None:
+        hi = len(a)
+
+    while lo < hi:
+        mid = (lo + hi) // 2
+        if a[mid] > x:
+            lo = mid + 1
+        else:
+            hi = mid
+
+    return lo
+
+
+class SequencePaginator(object):
+    def __init__(self, data, reverse=False):
+        self.scores, self.values = map(
+            list,
+            zip(*sorted(data, reverse=reverse)),
+        ) if data else ([], [])
+        self.reverse = reverse
+        self.search = functools.partial(
+            bisect.bisect_left if not reverse else reverse_bisect_left,
+            self.scores,
+        )
+
+    def get_result(self, limit, cursor=None):
+        if cursor is None:
+            cursor_score = None
```

Comment:
I didn't just instantiate a cursor here because the cursor requires a integer value for the first parameter and 0 is potentially a valid value in many cases. Explicitly using `None` as a sentinel value avoids that below.

This concept is called "value" in the `Cursor` constructor, but I've called it "score" here because "value" is ambiguous whether it's the score/key or the item, and I kept getting confused when writing this originally, so I went with the less confusing option.

Reply:
I gave up and this now mirrors the other implementation: 77c965e9782899dfad303324cc97a2d70e0cc6e3